### PR TITLE
Draft changes for h2o-httpclient support pre rfc 8441 connect and rfc 9298 connect-udp

### PR DESCRIPTION
**Pull request by antoniofastly**
Would appreciate feedback on my current approach.

I'm not sure why connect-udp over http2 does not seem to be working yet, still debugging that / adding a test.

### CODE REVIEW

**Review by kazuho — COMMENTED** [review]
Thank you for working on this. I think we are on the right path, most changes look good!

I've left some design-level comments; please let me know what you think.

File: include/h2o/url.h [hunk 1]
```diff
@@ -32,8 +32,12 @@ typedef struct st_h2o_url_scheme_t {
     int is_ssl;
 } h2o_url_scheme_t;
 
-extern const h2o_url_scheme_t H2O_URL_SCHEME_HTTP, H2O_URL_SCHEME_HTTPS, H2O_URL_SCHEME_MASQUE;
+extern const h2o_url_scheme_t H2O_URL_SCHEME_HTTP, H2O_URL_SCHEME_HTTPS;
 extern const h2o_url_scheme_t H2O_URL_SCHEME_MASQUE;
+/**
+ * no :scheme for CONNECT over H2/H3 prior to the introduction of extended CONNECT RFC 8441 / RFC 9220
+ */
+extern const h2o_url_scheme_t H2O_URL_SCHEME_CONNECT_NONE;
```
[the remark this answers is not in the document]
Comment (antoniofastly) [reply include/h2o/url.h:40]: I think that's the case.  src/httpclient.c sets a valid scheme but hpack/qpack are hard coded to omit path and scheme.  There are several ways to make the omit conditional:
1. pass in a sentinel like H2O_URL_SCHEME_CONNECT_NONE (or H2O_URL_SCHEME_CONNECT_NULL, or H2O_URL_SCHEME_NULL)
2. set the url scheme to NULL

Setting the scheme to NULL seems dangerous seems that's what you would get by default on a 0 initialized url struct you get from calloc.  Code that is not handling CONNECT will likely crash when it sees a NULL scheme.

File: include/h2o/httpclient.h [hunk 2]
```diff
@@ -257,6 +257,14 @@ struct st_h2o_httpclient_t {
...
      */
     const char *upgrade_to;
+    /**
+     * Protocol to use to create extended CONNECT and CONNECT-UDP tunnels.
+     * Standards:
+     *   RFC 8441 Extended Connect and websockets over HTTP/2
+     *   RFC 9220 Extended Connect and websockets over HTTP/3
+     *   RFC 9298 CONNECT-UDP over extended Connect
+     */
+    const char* extended_connect_protocol;
```
[the remark this answers is not in the document]
Comment (antoniofastly) [reply include/h2o/httpclient.h:267]: I can try it.  I'll send you a separate PR that implements that alternative.

I will need to look carefully at places that handle the upgrade_to header today and add Upgrade / Connection headers based on its presence.

File: include/h2o/url.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (kazuho) [reply include/h2o/url.h:40]: I agree that we should not be setting `scheme` to NULL.

What I am wondering if is if we could use `protocol.base == NULL` (or maybe the upgrade token being set to `h2o_httpclient_upgrade_to_connect`) as an indication that the request is traditional CONNECT.

WDYT?

[the remark this answers is not in the document]
Comment (antoniofastly) [reply include/h2o/url.h:40]: CONNECT requests with `protocol.base == NULL` seems like a good option.  I'll check for that in the new PR.

I think that access qpack does not have access to the upgrade_to field stored in the client which is not passed in.

**Comment by antoniofastly** [comment]
Closing in favor of https://github.com/h2o/h2o/pull/3307

File: src/httpclient.c [hunk 3]
```diff
@@ -73,6 +74,8 @@ static uint64_t io_timeout = DEFAULT_IO_TIMEOUT;
 static int ssl_verify_none = 0;
 static int exit_failure_on_http_errors = 0;
 static int program_exit_status = EXIT_SUCCESS;
+// Use draft03 syntax if false.
+static int connect_udp_is_rfc9298 = 0;
```
[the remark this answers is not in the document]
Comment (antoniofastly) [reply src/httpclient.c:78]: Addressed in https://github.com/h2o/h2o/pull/3307

File: lib/http2/hpack.c [hunk 4]
```diff
@@ -1013,31 +1013,37 @@ static void fixup_frame_headers(h2o_buffer_t **buf, size_t start_at, uint8_t typ
 
 void h2o_hpack_flatten_request(h2o_buffer_t **buf, h2o_hpack_header_table_t *header_table, uint32_t hpack_capacity,
                                uint32_t stream_id, size_t max_frame_size, h2o_iovec_t method, h2o_url_t *url,
-                               const h2o_header_t *headers, size_t num_headers, int is_end_stream)
+                               h2o_iovec_t protocol, const h2o_header_t *headers, size_t num_headers, int is_end_stream)
 {
-    int is_connect = h2o_memis(method.base, method.len, H2O_STRLIT("CONNECT"));
+    // Pre RFC8441 CONNECT over H2 has no path or scheme.
+    int has_scheme_and_path =
+        (url->scheme != &H2O_URL_SCHEME_CONNECT_NONE || !h2o_memis(method.base, method.len, H2O_STRLIT("CONNECT")));
```
[the remark this answers is not in the document]
Comment (antoniofastly) [reply lib/http2/hpack.c:1020]: Done

File: include/h2o/url.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (antoniofastly) [reply include/h2o/url.h:40]: Done in https://github.com/h2o/h2o/pull/3307

File: include/h2o/httpclient.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (antoniofastly) [reply include/h2o/httpclient.h:267]: See https://github.com/h2o/h2o/pull/3307

File: src/httpclient.c [hunk 5]
```diff
@@ -523,6 +567,9 @@ static void usage(const char *progname)
             "               verification\n"
             "  -X <local-udp-port>\n"
             "               specifies that the tunnel being created is a CONNECT-UDP tunnel\n"
+            "  --connect-udp-rfc-version <draft03 | rfc9298>\n"
+            "               specifies the spec version for the CONNECT-UDP tunnel: draft03 or\n"
+            "               rfc9298 (default: draft03)\n"
```
[the remark this answers is not in the document]
Comment (antoniofastly) [reply src/httpclient.c:572]: See https://github.com/h2o/h2o/pull/3307#discussion_r1387189524

**Pull request closed without merging**